### PR TITLE
Use a rails resource for courses-related routing

### PR DESCRIPTION
**Pull request by Bjvanminnen**
This PR aligns all of our courses related routing under a singular resource. The one exception is `/course/:course_name` which just does a redirect to `/courses/:course_name`. This is so that PLC courses can continue to use the /course route.

This also involved moving around some views/actions:

courses#index became courses#show (this is for the route /courses/:course_name)
home#courses became courses#index (this is for the route /courses)

I also moved the recent_courses logic from the home controller to the user model.

The PR diff itself is likely easier to read if you look at it on a per-commit basis.

Tested on my localhost that /courses, /course/all-the-plc-things, /courses/csp routes all still work.

### CODE REVIEW

File: dashboard/app/controllers/courses_controller.rb [hunk 1]
```diff
@@ -1,6 +1,14 @@
 class CoursesController < ApplicationController
   def index
-    course_name = params[:course].tr('-', '_').titleize
+    @recent_courses = current_user.try(:recent_courses)
+    @is_teacher = !!(current_user && current_user.teacher?)
```
Comment (aoby) [inline dashboard/app/controllers/courses_controller.rb]: how about `@is_teacher = !!current_user.try(:teacher?)` ?

File: dashboard/app/models/user.rb [hunk 2]
```diff
@@ -884,6 +887,28 @@ def in_progress_and_completed_scripts
...
+      {
+        id: script_id,
+        script_name: script_name,
+        courseName: data_t_suffix('script.name', script_name, 'title'),
+        description: data_t_suffix('script.name', script_name, 'description_short'),
+        # TODO: This can probably be generated on the client. If it can, we can also
+        # get rid of include Rails.application.routes.url_helpers
+        link: script_url(script_id),
+        image: "",
+        assignedSections: []
```
Comment (aoby) [inline dashboard/app/models/user.rb]: Are `image` and `assignedSections` needed? Why are they always blank and empty respectively? Are they filled in somewhere else?

File: dashboard/app/controllers/courses_controller.rb (hunk 1 above)
Comment (Bjvanminnen) [reply dashboard/app/controllers/courses_controller.rb]: agreed. i was going to change this in a future PR, but wanted this one to mostly just be moving stuff

File: dashboard/app/models/user.rb (hunk 2 above)
Comment (Bjvanminnen) [reply dashboard/app/models/user.rb]: They are used by React components (which will presumably warn if they're empty). I expect the shape of this data to still change a bit in the nearish future.

**Comment by Bjvanminnen** [comment]
This got (unintentionally) merged as part of https://github.com/code-dot-org/code-dot-org/pull/15133. I should have merged this PR, and then that one.

**Pull request closed without merging**